packages/cli/src/controller/publish-controller.ts

```diff
@@ -3,66 +3,102 @@
 
   if (!manifestIsV0_2_0(manifest)) {
     throw new Error('Unsupported project manifest spec, only 0.2.0 is supported');
   }
-
-  const deployment = await replaceFileReferences(ipfs, projectDir, manifest);
+  let ipfs: IPFSHTTPClient;
+  if (ipfsEndpoint) {
+    ipfs = IPFS.create({url: ipfsEndpoint});
+  }
+  const cluster = new Cluster(IPFS_CLUSTER_ENDPOINT, {
+    // optional custom headers for e.g. auth
+    headers: {Authorization: `Bearer ${authToken}`},
+  });
+  const deployment = await replaceFileReferences(projectDir, manifest, cluster, ipfs);
 
   // Upload schema
-  return uploadFile(ipfs, deployment.toDeployment());
+  return uploadFile(deployment.toDeployment(), cluster, ipfs);
 }
 
 /* Recursively finds all FileReferences in an object and replaces the files with IPFS references */
-async function replaceFileReferences<T>(ipfs: IPFS.IPFSHTTPClient, projectDir: string, input: T): Promise<T> {
+async function replaceFileReferences<T>(
+  projectDir: string,
+  input: T,
+  cluster: Cluster,
+  ipfs?: IPFS.IPFSHTTPClient
+): Promise<T> {
   if (Array.isArray(input)) {
-    return (await Promise.all(input.map((val) => replaceFileReferences(ipfs, projectDir, val)))) as unknown as T;
+    return (await Promise.all(
+      input.map((val) => replaceFileReferences(projectDir, val, cluster, ipfs))
+    )) as unknown as T;
   } else if (typeof input === 'object') {
     if (input instanceof Map) {
       input = mapToObject(input) as T;
     }
 
     if (isFileReference(input)) {
-      input.file = await uploadFile(ipfs, fs.createReadStream(path.resolve(projectDir, input.file))).then(
+      input.file = await uploadFile(fs.createReadStream(path.resolve(projectDir, input.file)), cluster, ipfs).then(
         (cid) => `ipfs://${cid}`
       );
     }
 
     const keys = Object.keys(input) as unknown as (keyof T)[];
     await Promise.all(
       keys.map(async (key) => {
-        input[key] = await replaceFileReferences(ipfs, projectDir, input[key]);
+        input[key] = await replaceFileReferences(projectDir, input[key], cluster, ipfs);
       })
     );
   }
 
   return input;
 }
 
-async function uploadFile(ipfs: IPFS.IPFSHTTPClient, content: FileObject | FileContent): Promise<string> {
-  const result = await ipfs.add(content, {pin: true, cidVersion: 0});
-  return result.cid.toString();
+export async function uploadFile(
+  content: string | fs.ReadStream,
+  cluster: Cluster,
+  ipfs?: IPFS.IPFSHTTPClient
+): Promise<string> {
+  const clientPromise = ipfs
+    ? ipfs.add(content, {pin: true, cidVersion: 0})
+    : Promise.reject(new Error('Ipfs server not provided'));
+  let contentBuffer: Uint8Array;
+  if (determineStringOrFsStream(content)) {
+    contentBuffer = new Blob([fs.readFileSync(content.path)]);
+  } else {
+    contentBuffer = new Blob([content]);
+  }
+  const clusterPromise = cluster.add(contentBuffer, {cidVersion: 0, rawLeaves: false});
+  const results = await Promise.allSettled([clientPromise, clusterPromise]);
```

Comment:
This might be better suited https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/any